pkg/pillar/vault/key.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2022 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package vault
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	etpm "github.com/lf-edge/eve/pkg/pillar/evetpm"
+)
+
+var (
+	errInvalidKeyLen = errors.New("unexpected key length")
+)
+
+const (
+	vaultKeyLen = 32 //bytes
+)
+
+func retrieveTpmKey(log *base.LogObject, useSealedKey bool) ([]byte, error) {
+	if useSealedKey {
+		return etpm.FetchSealedVaultKey(log)
+	}
+	return etpm.FetchVaultKey(log)
+}
+
+// retrieveCloudKey is to support pre-5.6.2 devices, remove once devices move to 5.6.2
+func retrieveCloudKey() ([]byte, error) {
+	// For now, return a dummy key, until controller support is ready.
+	cloudKey := []byte("foobarfoobarfoobarfoobarfoobarfo")
+	return cloudKey, nil
+}
+
+// cloudKeyOnlyMode is set when the key is used only from cloud, and not from TPM.
+func deriveVaultKey(log *base.LogObject, cloudKeyOnlyMode, useSealedKey, tpmKeyOnlyMode bool) ([]byte, error) {
+	// First fetch Cloud Key
+	cloudKey, err := retrieveCloudKey()
+	if err != nil {
+		return nil, err
+	}
+	// For pre 5.6.2 devices, remove once devices move to 5.6.2
+	if cloudKeyOnlyMode {
+		log.Functionf("Using cloud key")
+		return cloudKey, nil
+	}
+	tpmKey, err := retrieveTpmKey(log, useSealedKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if tpmKeyOnlyMode == false {
+		log.Notice("Calling mergeKeys")
+		return mergeKeys(log, tpmKey, cloudKey)
+	}
+	log.Notice("Using TPM key only")
+	return tpmKey, nil
+}
+
+// stageKey is responsible for talking to TPM and Controller
+// and preparing the key for accessing the vault
+// returns function to unstage the key
+func stageKey(log *base.LogObject, cloudKeyOnlyMode, useSealedKey, tpmKeyOnlyMode bool, keyDirName string, keyFileName string) (func(), error) {
+	// Create a tmpfs file to pass the secret to fscrypt
+	if err := os.MkdirAll(keyDirName, 755); err != nil {
+		return nil, fmt.Errorf("error creating keyDir %s %v", keyDirName, err)
+	}
+
+	if _, _, err := execCmd("mount", "-t", "tmpfs", "tmpfs", keyDirName); err != nil {
```

Comment:
Can we use base package for Exec() calls in this package?

Reply:
There might even be an os.Mount() call.
But I wouldn't change that as part of this PR; better to keep the structural changes separate from other desirable changes.